---
title: "DESeq2_2F_Batch"
author: "RuiLi"
date: "07/24/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# source("https://bioconductor.org/biocLite.R")
# biocLite("DESeq2")
# biocLite("apeglm")
library(DESeq2)
library('apeglm')
thresh_p <- 0.05
thresh_FC <- 1
```

## Read Data
```{r}
df <- read.table('counts.gene_id.txt', sep="\t",
                 header=TRUE, row.names=1)

cts <- df[,6:ncol(df)]

colnames(cts) <- gsub("\\.Aligned.out.bam$", "", colnames(cts))
colnames(cts) <- gsub("...star.", "", colnames(cts))
colnames(cts) <- gsub("C241", "C261", colnames(cts)) # fix C241, C261 misname

cts <- cts[, c(1:8, 10:11, 9, 12)]  # fix C241, C261 misname

cts <- cts[, c(1,5,9,
               2,6,10,
               3,7,11,
               4,8,12)]  # fix order

cts <- as.matrix(cts)
dim(cts)
head(cts)

## Filter Based on Expression
exp_idx <- rowSums(cts) >= 0  # default 10
cts <- cts[exp_idx, ]
dim(cts)
head(cts)
```


## DESeq Experiment Design
```{r}
type <- factor(rep(c('yhp1','yox1','yox1yhp1','WT'), each=3))
batch <- factor(rep(c('A','B','C'), 4))
coldata <- data.frame(row.names=colnames(cts), type, batch)
coldata
```


## Create dds obj
```{r}
dds <- DESeqDataSetFromMatrix(countData = cts, 
                              colData = coldata, 
                              design = ~ batch + type )  # converted to alph-order
dds$type <- relevel(dds$type, ref = "WT")
dds
dds <-DESeq(dds)
resultsNames(dds)

# res <- results(dds, name="type_yhp1_vs_WT", alpha = thresh_p) 
# name <- 'type_yhp1_vs_WT_default' 

res1 <- lfcShrink(dds, coef="type_yhp1_vs_WT", type="apeglm")
name1 <- 'type_yhp1_vs_WT'

res2 <- lfcShrink(dds, coef="type_yox1_vs_WT", type="apeglm")
name2 <- 'type_yox1_vs_WT'

res3 <- lfcShrink(dds, coef="type_yox1yhp1_vs_WT", type="apeglm")
name3 <- 'type_yox1yhp1_vs_WT'
```
```{r}
dds <- DESeqDataSetFromMatrix(countData = cts, 
                              colData = coldata, 
                              design = ~ batch + type )  # converted to alph-order
dds$type <- relevel(dds$type, ref = "yox1yhp1")
dds
dds <-DESeq(dds)
resultsNames(dds)

# res <- results(dds, name="type_yhp1_vs_WT", alpha = thresh_p) 
# name <- 'type_yhp1_vs_WT_default' 

res4 <- lfcShrink(dds, coef="type_yhp1_vs_yox1yhp1", type="apeglm")
name4 <- 'type_yhp1_vs_yox1yhp1'

res5 <- lfcShrink(dds, coef="type_yox1_vs_yox1yhp1", type="apeglm")
name5 <- 'type_yox1_vs_yox1yhp1'
```

## Normal: WT as Ref
```{r}
## Change
res <- res1 #fly
name <- name1 #fly

res <- res2 #fly
name <- name2 #fly

res <- res3 #fly
name <- name3 #fly

res <- res4 #fly
name <- name4 #fly

res <- res5 #fly
name <- name5 #fly
```

```{r}
## Contrast specific Summary
sink(paste(name, 'summary.txt', sep = '.'))
name
summary(res)
table(res$padj<thresh_p)
sink()

## Order by adjusted p-value
res <- res[order(res$padj), ]
## Merge with normalized count data
resdata <- merge(as.data.frame(res), as.data.frame(counts(dds, normalized=TRUE)), 
                 by="row.names", sort=FALSE)
names(resdata)[1] <- "Gene"
head(resdata)
## Write results
write.csv(resdata, row.names = FALSE, 
          file=paste(name, 'deseq.csv', sep = '.'))
```

```{r}
## Contrast Specific Plots

## Examine plot of p-values
png(paste(name, "pvalues.png", sep = '.'), 1000, 1000, pointsize=20)
hist(res$pvalue, breaks=50, col="grey", 
     main = paste('Histogram of p-values (un-adjusted)', name, sep = "\n"), 
     xlab = 'pvalues', ylab = 'Frequency')
dev.off()

png(paste(name, "padj.png", sep = '.'), 1000, 1000, pointsize=20)
hist(res$padj, breaks=50, col="grey", 
     main = paste('Histogram of p-values (adjusted)', name, sep = "\n"), 
     xlab = 'pvalues', ylab = 'Frequency')
dev.off()


## MA plot
## Could do with built-in DESeq2 function:
## DESeq2::plotMA(dds, ylim=c(-1,1), cex=1)
## I like mine better:
maplot <- function (res, thresh=thresh_p, labelsig=FALSE, textcx=1, ...) {
  with(res, plot(baseMean, log2FoldChange, pch=20, cex=.5, log="x", ...))
  with(subset(res, padj<thresh_p), points(baseMean, log2FoldChange, col="red", pch=20, cex=1.5))
  if (labelsig) {
    require(calibrate)
    with(subset(res, padj<thresh_p), textxy(baseMean, log2FoldChange, labs=Gene, cex=textcx, col=2))
  }
}
png(paste(name, "MAplot.png", sep="."), 1500, 1000, pointsize=20)
maplot(resdata, main=paste("MAplot", name, sep="\n"))
dev.off()


## Volcano plot with "significant" genes labeled
volcanoplot <- function (res, lfcthresh=2, sigthresh=thresh_p, 
                         main=paste("Volcano Plot", name, sep='\n'), 
                         legendpos="topright", labelsig=FALSE, textcx=1, ...) {
  with(res, plot(log2FoldChange, -log10(pvalue), pch=20, main=main, ...))
  with(subset(res, padj<sigthresh ), points(log2FoldChange, -log10(pvalue), pch=20, col="orange", ...))
  with(subset(res, abs(log2FoldChange)>lfcthresh), points(log2FoldChange, -log10(pvalue), pch=20, col="green", ...))
  with(subset(res, padj<sigthresh & abs(log2FoldChange)>lfcthresh), points(log2FoldChange, -log10(pvalue), pch=20, col="red", ...))
  if (labelsig) {
    require(calibrate)
    with(subset(res, padj<sigthresh & abs(log2FoldChange)>lfcthresh), textxy(log2FoldChange, -log10(pvalue), labs=Gene, cex=textcx, ...))
  }
  legend(legendpos, xjust=1, yjust=0.5, 
         legend=c(paste("FDR<",sigthresh,sep=""), 
                  paste("|LogFC|>",lfcthresh,sep=""), 
                  "both"), pch=20,
         col=c("orange","green","red"))
}
png(paste(name, "volcano.png", sep='.'), 1200, 1000, pointsize=20)
volcanoplot(resdata, lfcthresh=thresh_FC, sigthresh=thresh_p, textcx=.8, xlim=c(-2.3, 2))
dev.off()

```



## Overall Plot
```{r}
# Regularized log transformation for clustering/heatmaps, etc
vsd <- vst(dds, blind=FALSE)
rld <- rlog(dds, blind=FALSE)
head(assay(vsd))
hist(assay(rld))
hist(assay(vsd))



# Plot dispersions
png("qc-dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds, main="Dispersion plot")
dev.off()


## PCA
png("qc-PCA-samples.batch.png", w=1000, h=1000, pointsize=20)
plotPCA(vsd, intgroup=c("batch"))
dev.off()

png("qc-PCA-samples.type.png", w=1000, h=1000, pointsize=20)
plotPCA(vsd, intgroup=c("type"))
dev.off()


## Distance
sampleDists <- dist(t(assay(vsd)))
library("RColorBrewer")
library('pheatmap')
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, sep="-")  # change factor name
colnames(sampleDistMatrix) <- paste(vsd$condition, sep="-")
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
png(paste("qc-heatmap-samples", name, "png", sep = '.'), w=1000, h=1000, pointsize=20)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
dev.off()

## Official Heatmap of Counts
library("pheatmap")
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)[,c("type","batch")])
png("qc-heatmap-counts.png", w=1000, h=1000, pointsize=20)
pheatmap(assay(dds)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
dev.off()
```



















## Merge Results
```{r}
tab1 = read.csv(paste(name1, 'deseq.csv', sep='.'))[, 1:6]
tab2 = read.csv(paste(name2, 'deseq.csv', sep='.'))[, 1:6]
tab3 = read.csv(paste(name3, 'deseq.csv', sep='.'))[, 1:6]
tab4 = read.csv(paste(name4, 'deseq.csv', sep='.'))[, 1:6]
tab5 = read.csv(paste(name5, 'deseq.csv', sep='.'))[, 1:6]

names(tab1)[3:6] = paste(names(tab1)[3:6], name1, sep='.')
names(tab2)[3:6] = paste(names(tab2)[3:6], name2, sep='.')
names(tab3)[3:6] = paste(names(tab3)[3:6], name3, sep='.')
names(tab4)[3:6] = paste(names(tab4)[3:6], name4, sep='.')
names(tab5)[3:6] = paste(names(tab5)[3:6], name4, sep='.')

tab_exp = read.csv(paste(name1, 'deseq.csv', sep='.'))
tab_exp = tab_exp[, c(1,2, 7:dim(tab_exp)[2])]

require(plyr)
tab_merged = join_all(
  list(tab1, tab2, tab3, tab4, tab_exp), 
  by=c('Gene', 'baseMean'))

## Write results
write.csv(tab_merged, row.names = FALSE, 
          file=paste('merged', 'deseq.csv', sep = '.'))

write.table(tab_merged[,1], 
            row.names = FALSE, col.names = FALSE,
            quote = FALSE,
            file='gene_id.txt')

```




## Annotate Gene_ID

# ```{r}
# ## Panther Anno of Ensembl_ID (multi wrong mapping)
# tab_deseq = read.csv('merged.deseq.csv')
# 
# ## read Panther Anno (from website)
# tab_anno = read.csv('pantherGeneList.txt', 
#                     sep = '\t', 
#                     header = FALSE)[, 1:5]  # remove species column
# names(tab_anno) = c('PantherID', 'MappedID',
#                     "GeneName","PantherFamily", 
#                     "PantherProteinClass")
# dup_row = grep(',', tab_anno$MappedID) #multi-mapped mapped_id
# 
# for (row_num in dup_row){
#   old_row = tab_anno[row_num,]
#   ids_str = toString(old_row$MappedID)
#   ids = as.list(strsplit(ids_str, ",")[[1]])
#   for (id in ids){
#     new_row = old_row
#     new_row$MappedID = id
#     tab_anno = rbind(tab_anno, new_row)
#   }
# }
# tab_anno = tab_anno[-dup_row,]  #remove old_rows
# 
# merged = merge(tab_deseq, tab_anno, 
#                by.x='Gene', by.y='MappedID',
#                all.x=TRUE, all.y=FALSE,
#                all = L,
#                sort = TRUE,
#                no.dups= TRUE
#                )
# 
# ## Write results
# write.csv(merged, row.names = FALSE, 
#           file=paste('merged', 'deseq.panther.csv', sep = '.'))
# ```

# ```{r}
# ## ensembldb (no v92)
# biocLite('EnsDb.Mmusculus.v79')
# library(EnsDb.Mmusculus.v79)
# 
# ## Making a "short cut"
# edb <- EnsDb.Hsapiens.v86
# ## print some informations for this package
# edb
# tab_deseq = read.csv('merged.deseq.csv')[1:100, ]
# 
# ```


```{r}
library('refGenome')
ens <- ensemblGenome()
basedir(ens) <- system.file("extdata", package="refGenome")
# ens_gtf <- "Mus_musculus.GRCm38.92.gtf"
# read.gtf(ens, ens_gtf)
# saveRDS(ens, 'Mus_musculus.GRCm38.92.rds')

ens_rdata <- 'Mus_musculus.GRCm38.92.rds'
ens = readRDS(paste(basedir(ens), ens_rdata, sep = '/'))

# hit<-extractByGeneId(ens,"ENSMUSG00000096730")
# hit<-extractByGeneName(ens,"Sra1")
# getGeneTable(hit)$gene_name
# getGeneTable(hit)$gene_id

gene_ids = getGeneTable(ens)$gene_id
gene_names = getGeneTable(ens)$gene_name
tab_anno = data.frame(gene_ids, gene_names)

tab_deseq = read.csv('merged.deseq.csv')

tab_merged = merge(tab_deseq, tab_anno,
                   by.x = 'Gene', by.y = 'gene_ids', 
                   )
# fix name
names(tab_merged)[1] = 'ensembl_id'
# fix order
d2 = dim(tab_merged)[2]
tab_merged = tab_merged[,c(1, d2, 2:(d2-1))]

write.csv(tab_merged, 'deseq.anno.csv', row.names = FALSE)
```

```{r}
## significant genes
tab = read.csv('./final/20180607.deseq.anno.csv')
# 
# library(xlsx)
# write.xlsx(tab, '20180607.deseq.anno.xlsx')

stat_sig1 = tab$pvalue.Normal_WT_vs_KO < thresh_p 
bio_sig1 = abs(tab$log2FoldChange.Normal_WT_vs_KO) > thresh_FC
sig1 = stat_sig1 & bio_sig1
write.csv(
  tab[sig1,], 
  paste('Normal_WT_vs_KO', 'sig.csv', sep='.'),
  row.names = FALSE)

stat_sig2 = tab$pvalue.Deprived_WT_vs_KO < thresh_p 
bio_sig2 = abs(tab$log2FoldChange.Deprived_WT_vs_KO) > thresh_FC
sig2 = stat_sig2 & bio_sig2
write.csv(
  tab[sig2,], 
  paste('Deprived_WT_vs_KO', 'sig.csv', sep='.'),
  row.names = FALSE)

stat_sig3 = tab$pvalue.WT_Normal_vs_Deprived < thresh_p 
bio_sig3 = abs(tab$log2FoldChange.WT_Normal_vs_Deprived) > thresh_FC
sig3 = stat_sig3 & bio_sig3
write.csv(
  tab[sig3,], 
  paste('WT_Normal_vs_Deprived', 'sig.csv', sep='.'),
  row.names = FALSE)

stat_sig4 = tab$pvalue.KO_Normal_vs_Deprived < thresh_p 
bio_sig4 = abs(tab$log2FoldChange.KO_Normal_vs_Deprived) > thresh_FC
sig4 = stat_sig4 & bio_sig4
write.csv(
  tab[sig4,], 
  paste('KO_Normal_vs_Deprived', 'sig.csv', sep='.'),
  row.names = FALSE)

```

