---
title: "DEseq2 Template"
author: "RuiLi"
date: "04/10/2018"
output:
  html_document:
    toc: yes
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
---
# Updates
- ylim for VolcanoPlot: (0,30)
- add TPM normalization (with scater)
- add normalization options in process_deseq_res
- add annotation in pipeline, just provide an annotation table with V1 as gene_id

# Communication
- For Namgyu in Dohoon lab
- See MDA_MB231_vs_MCF10A.deseq2.xlsx for DE results
    - see padj as the FDR of differential expression
    - see LFC as the Log2FoldChange
    - see the last 9 columns are TPM for 9 samples

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(ggplot2)
library("scater")
library(WriteXLS)


thresh_p <- 0.05
thresh_LFC <- 1  # 0.585


volcanoplot <- function (res, 
                         lfcthresh=thresh_LFC, sigthresh=thresh_p, 
                         name='name', 
                         legendpos="topright", labelsig=FALSE, textcx=1, ...) {
    main=paste("Volcano Plot (FDR, LFC)", name, sep='\n')
    with(res, 
         plot(log2FoldChange, -log10(padj), 
              pch=20, main=main, 
              ylim=c(0,30),
              ...))
    with(subset(res, padj<sigthresh ), 
         points(log2FoldChange, -log10(padj), pch=20, col="orange", ...))
    with(subset(res, abs(log2FoldChange)>lfcthresh), 
         points(log2FoldChange, -log10(padj), pch=20, col="green", ...))
    with(subset(res, padj<sigthresh & abs(log2FoldChange)>lfcthresh), 
         points(log2FoldChange, -log10(padj), pch=20, col="red", ...))
    if (labelsig) {
        require(calibrate)
        with(subset(res, padj<sigthresh & abs(log2FoldChange)>lfcthresh), 
             textxy(log2FoldChange, -log10(padj), labs=Gene, cex=textcx, ...))
    }
    legend(legendpos, xjust=1, yjust=0.5, 
           legend=c(paste("FDR<",sigthresh,sep=""), 
                    paste("|LogFC|>",lfcthresh,sep=""), 
                    "both"), pch=20,
           col=c("orange","green","red"))
}

maplot <- function (res, thresh=thresh_p, labelsig=FALSE, textcx=1, ...) {
  with(res, 
       plot(baseMean, log2FoldChange, pch=20, cex=.5, log="x", ...))
  with(subset(res, padj<thresh_p), 
       points(baseMean, log2FoldChange, col="red", pch=20, cex=1.5))
  if (labelsig) {
    require(calibrate)
    with(subset(res, padj<thresh_p), 
         textxy(baseMean, log2FoldChange, labs=Gene, cex=textcx, col=2))
  }
}


process_deseq_res <- function(res, name='name', anno="df_anno", normalized="df_tpm"){
    ## Summary
    print(name)
    summary(res, alpha=thresh_p)
    table(res$padj<thresh_p)
    
    ## Output
    ## Order by LFC (up to down)
    res <- res[order(res$log2FoldChange, decreasing = T), ]
    ## Merge with normalized count data
    resdata <- merge(as.data.frame(res), 
                     # as.data.frame(counts(dds, normalized=TRUE)), 
                     normalized,
                     by="row.names", sort=FALSE)
    names(resdata)[1] <- "Gene"
    resdata <- merge(anno, resdata, 
                     all.x=F, all.y=T,
                     by.x=1, by.y=1)
    head(resdata)
    ## Write results
    # write.csv(resdata, row.names = FALSE, 
    #           file=paste(name, 'deseq2.csv', sep = '.'))
    WriteXLS(x = resdata, 
         ExcelFileName = paste(name, 'deseq2.xlsx', sep = '.'), row.names = F, SheetNames = 'sheet1', na = '-')
    
    ## Hist of P-values/FDR
    ## Examine plot of p-values
    hist(res$pvalue, breaks=50, col="grey", 
         main = paste('Histogram of p-values (un-adjusted)', name, sep = "\n"), 
         xlab = 'pvalues', ylab = 'Frequency')
    
    hist(res$padj, breaks=50, col="grey", 
         main = paste('Histogram of p-values (adjusted)', name, sep = "\n"), 
         xlab = 'padj', ylab = 'Frequency')
    
    maplot(resdata, main=paste("MAplot", name, sep="\n"))
    volcanoplot(resdata, 
                lfcthresh=thresh_LFC, sigthresh=thresh_p, 
                textcx=.8, xlim=c(-2.3, 2), name=name)
}
```


# Read Data
```{r readdata}
df <- read.table('../GSE75168_MCF10A_MCF7_MDA-MB-231_HTSeq_Counts.txt.gz', 
                 sep="\t", header=TRUE, 
                 row.names = 1) # row.name in cts(matrix)
head(df)
dim(df)
```

## Read length and annotation
```{r}
# anno <- read.table("~/github/bpipes/rnaseq/de/annotations/gencode.v29.primary_assembly.anno.txt.gz", header=T)
anno <- read.table("~/github/bpipes/rnaseq/de/annotations/gencode.hu.v21/gencode.v21.primary_assembly.anno.txt.gz", header=T)
head(anno)
dim(anno)
```

## Annotate
```{r}
# df$Ensembl_ID <- gsub("\\..*$", "", df$Ensembl_ID) # remove tr_id
# anno$Gene <- gsub("\\..*$", "", anno$Gene)
df <- merge(anno, df, by.x=1, by.y=0)
dim(df)
```

## get cts.matrix for DESeq2
```{r}
cts <- df[,9:ncol(df)]
# colnames(cts) <- gsub("\\.Aligned.sortedByCoord.out.bam$", "", colnames(cts))
# colnames(cts) <- gsub("...star.", "", colnames(cts))
# colnames(cts) <- gsub("_S\\d{,2}_L00\\d?_R1_001.fastq.gz", "", colnames(cts))
# colnames(cts)

# colname_formatted <- c("KO_D0_3", "WT_D0_3", "KO_D2_3", "WT_D2_3", "KO_D8_3", "WT_D8_3",
#                    "KO_D0_1", "KO_D0_2", "KO_D2_1", "KO_D2_2", "KO_D8_1", "KO_D8_2", 
#                    "WT_D0_1", "WT_D0_2", "WT_D2_1", "WT_D2_2", "WT_D8_1", "WT_D8_2")
# paste(colnames(cts), colname_formatted, sep = "->")
# colnames(cts) <- colname_formatted
# cts <- cts[, sort(colname_formatted)]

## raw matrix
cts <- as.matrix(cts)
row.names(cts) <- df$Gene
```

## Filter Based on Expression
```{r}
expression_filter <- rowSums(cts) >= 1  # default 10
cts <- cts[expression_filter, ]
df <- df[expression_filter, ]
dim(cts)
dim(df)
```

# TPM calculation
```{r}
tpm <- calculateTPM(cts, df$Length)
tpm <- data.frame(tpm)
colnames(tpm) <- paste("TPM", colnames(tpm), sep = ":")
tpm_out <- merge(anno, tpm, by.x=1, by.y=0)
WriteXLS(x = tpm_out, 
         ExcelFileName = 'TPM.xlsx', row.names = F, SheetNames = 'sheet1', na = '-')
```


# DESeq Experiment Design
- input: cts

## subset of data
```{r}
cts_ <- cts[, c(1,2,3,7,8,9)]
```



```{r design}
type <- factor(rep(c("MCF10A","MDA_MB231"), each=3))
batch <- factor(rep(rep(c('R1','R2','R3'), each=1), 2))
# group <- factor(paste(genotype, time, sep = "_"))
coldata <- data.frame(row.names=colnames(cts_), type, batch)
coldata
```

## Model fitting
```{r}
dds <- DESeqDataSetFromMatrix(countData = cts_, 
                              colData = coldata, 
                              design = ~ type )  # converted to alph-order
dds$type <- relevel(dds$type, ref = "MCF10A")
dds$batch <- relevel(dds$batch, ref = "R1")

dds
dds <-DESeq(dds)
deseq_normalized <- as.data.frame(counts(dds, normalized=TRUE))
resultsNames(dds)
```


# QC Plots

## Data transformation
```{r}
vsd <- vst(dds, blind=FALSE)

#rld <- rlog(dds, blind=FALSE)
counts <- counts(dds, normalized=0)
logCounts <- log10(counts +1 )

normed <- counts(dds, normalized=1)
logNormed <- log10(normed+1)
```

## Histogram of normalized data
```{r}
hist(logNormed, main='log10(x+1) normalized expression')
```

## Dispersion plot
```{r}
plotDispEsts(dds, main="Dispersion plot")
```


## PCA plots
- look at batch effect 
```{r}
pcaData <- plotPCA(vsd, intgroup=c("type", 'batch'), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2, color=batch, shape=type)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

## Heatmap
- samples clustered by time, then by batch
```{r}
library("RColorBrewer")
library('pheatmap')
library("PoiClaClu")
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
poisd <- PoissonDistance(t(counts(dds)))
samplePoisDistMatrix <- as.matrix( poisd$dd ) 
rownames(samplePoisDistMatrix) <- paste( colnames(dds), sep="-" ) 
colnames(samplePoisDistMatrix) <- NULL 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows=poisd$dd,
         clustering_distance_cols=poisd$dd,
         col=colors, 
         clustering_method='complete'
        )

```

# Results from different contrasts

## type_MDA_MB231_vs_MCF10A
```{r}
res <- lfcShrink(dds, coef = "type_MDA_MB231_vs_MCF10A", type = 'ashr')
name <- "MDA_MB231_vs_MCF10A"
process_deseq_res(res = res, name = name, anno=anno, normalized=tpm)

```


<!-- ## Simple Contrast -->
<!-- ```{r} -->
<!-- res <- lfcShrink(dds, contrast = c('type', 'KO_D0', 'WT_D0'), type = 'ashr') -->
<!-- name <- "D0_KO_vs_WT" -->
<!-- process_deseq_res(res = res, name = name) -->
<!-- ``` -->


<!-- ## Complex Contrast -->
<!-- - KO_vs_WT_in_D8_vs_D2 -->
<!-- - (K8 - W0) - (K2 - W0) - (W8 - W0) + (W2 - W0) -->
<!-- ```{r} -->
<!-- res <- lfcShrink(dds, -->
<!--                  contrast = c(0, #Intercept -->
<!--                               0, #type_KO_D0_vs_WT_D0 -->
<!--                               -1, #type_KO_D2_vs_WT_D0 -->
<!--                               1, #type_KO_D8_vs_WT_D0 -->
<!--                               1, #type_WT_D2_vs_WT_D0 -->
<!--                               -1, #type_WT_D8_vs_WT_D0 -->
<!--                               0, #batch_Nov_vs_Oct -->
<!--                               0 -->
<!--                               ), -->
<!--                  type = 'ashr') -->
<!-- name <- "KO_vs_WT_in_D8_vs_D2" -->
<!-- process_deseq_res(res = res, name = name) -->
<!-- ``` -->

