---
title: "TPM from Count: R is not accurate"
author: "RuiLi"
date: "7/26/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Read Count Table
Read, fix order and fix names, same as DESeq2, but keep metadata
```{r}
df <- read.table('../counts.gene_id.txt', sep="\t",
                 header=TRUE, row.names=1)

cts <- df[,6:ncol(df)]

colnames(cts) <- gsub("\\.Aligned.out.bam$", "", colnames(cts))
colnames(cts) <- gsub("...star.", "", colnames(cts))
colnames(cts) <- gsub("C241", "C261", colnames(cts)) # fix C241, C261 misname

cts <- cts[, c(1:8, 10:11, 9, 12)]  # fix C241, C261 misname

cts <- cts[, c(1,5,9,
               2,6,10,
               3,7,11,
               4,8,12)]  # fix order

cts <- as.matrix(cts)
dim(cts)
head(cts)
cst <- data.frame(cts)

df2 <- merge(df[,1:5], cts,
             by="row.names", 
             sort = 0)

colnames(df2)[1] <- "gene_id" 
```

## Calculate TPM (transcripts per million)
## Myself
```{r}
count <- df2[,7:18]
row.names(count) <- df2$gene_id

RPK <- count/(df2$Length/1000)
sum_RPK <- colSums(RPK)
TPM <- RPK/sum_RPK * 1000000

paste(colSums(TPM))
```

## Package
https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ 
```{r}
countToTpm <- function(counts, effLen)
{
    rate <- log(counts) - log(effLen)
    denom <- log(sum(exp(rate)))
    exp(rate - denom + log(1e6))
}
tpm <- countToTpm(count, df2$Length)
```

## Output
```{r}
# Paste
colnames(df2)[7:18] <- paste(colnames(df2)[7:18], "count", sep = '.')
colnames(TPM) <- paste(colnames(TPM), "TPM", sep = '.')

out_df <- merge(df2, TPM, 
      by.x='gene_id', by.y = 'row.names')

head(out_df)

write.csv(out_df, 'count_tpm.csv', row.names = FALSE)
# library(xlsx)
# write.xlsx(out_df, "merged.xlsx")
```
