---
title: "ChIPPeakAnno for Tessa"
author: "RuiLi"
date: "03/10/2019"
output:
  html_document:
    code_folding: hide
    toc: yes  # table of content
    toc_float: yes
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("ChIPpeakAnno")
library(WriteXLS)
require(dplyr)
library("GenomicFeatures")
```

## Notes:
1. **name.narrowPeak**: visualize peaks
2. name.anno.WithDup.xlsx: all peaks annotated, if multiple annotation for the same peak, multiple annotation lines are generated
3. name.anno.WithDup.protein_coding.xlsx: same as 2, but only protein_coding genes included in annotation. 
4. name.anno.collapsed.xlsx: same as 2, but each peak has only one line 
5. name.anno.protein_coding.collapsed.xlsx: same as 4, but only protein_coding genes included
5. **full_anno.xlsx**: same as 4, plus all unannotated peaks * (recommended)


## Code Log:
- 2019/07/30 developed for Merav ATAC-seq
- read anno.gr from genecode.gtf
- annotate on transcript level (TSS)
- collapse multiple-annotation by peak_id
- merge with original peak list to include unannotated peaks
- 2019/11/13
- added NarrowPeak output
- added xslx output
- 2020/03
- filter FDR
- Read GTF as TxDb
- Rui's precedence: moved intron after Immideate downstream
- Removed comma in peak info columns, lead to correct full_anno.xslx
- Added GO Enrichment and GO visualization with word cloud

## Parameters
```{r}
prefix <- 'SrcRepsBroad.Q0.05'
peakFile <- "../macs2_broad/SrcRepsBroad_peaks.broadPeak"
gtf_file <- "/project/umw_mccb/genome/Homo_sapiens/ucsc_hg38_primary/gencode.v29.primary_assembly.annotation.fixed.gtf"
max_fdr <- 0.05
```

## Prep AnnoData
### TxDb
- did not find an easy way to convert, also gencode29 used here
```{r}
# library(TxDb.Hsapiens.UCSC.hg38.knownGene)
# #library(TxDb.Mmusculus.UCSC.mm10.knownGene)
# TxDb <- TxDb.Hsapiens.UCSC.hg38.knownGene
# TxDb.gr <- toGRanges(TxDb)

library(rtracklayer)  # for the import() function
gtfgr <- import(gtf_file)
TxDb <- makeTxDbFromGRanges(gtfgr)
TxDb.gr <- toGRanges(TxDb)
TxDb
```

### genecode gene/transcript granges from GenCode.v29
```{r}
# install.packages("remotes")
# remotes::install_github("acidgenomics/freerange")
# library(freerange)
# geneCode.gene.gr <- makeGRangesFromGTF(gtf_file, 
#                                        level = c("genes"),
#                                        .checkAgainstTxDb = FALSE)
# 
# geneCode.tr.gr <- makeGRangesFromGTF(gtf_file, 
#                                      level = c("transcripts"),
#                                     .checkAgainstTxDb = FALSE)
# 
# saveRDS(geneCode.tr.gr, "../geneCode.tr.gr.RDS")
# saveRDS(geneCode.gene.gr, "../geneCode.gene.gr.RDS")

geneCode.gene.gr <- readRDS("/home/rl44w/github/bpipes/snakemake/chip_atac/postSnakemake/db/geneCode.gene.gr.RDS")  # skipped loading
geneCode.tr.gr <- readRDS("/home/rl44w/github/bpipes/snakemake/chip_atac/postSnakemake/db/geneCode.tr.gr.RDS")

#head(geneCode.tr.gr)
```

## Read peaks
```{r}
peaks.df <- read.table(peakFile, header = F)
dim(peaks.df)
colnames(peaks.df) <- c("seqnames", "start", "end", "peak_id", "score", "strand", "signalValue", "pValue", "qValue" )
peaks.df <- subset(peaks.df, qValue> -log10(max_fdr))
dim(peaks.df)
peaks.gr <- makeGRangesFromDataFrame(peaks.df, keep.extra.columns = T)
```

## Export Peaks to NarrowPeak format
```{r}
# df <- peaks.df
# bed <- data.frame(
#   seqnames=df$seqnames,
#   start=df$start,
#   end=df$end,
#   name=df$peak_id,
#   score=df$score,
#   strand=df$strand,
#   signalValue=df$Fold,
#   pValue=df$p.value,
#   qValue=df$FDR,
#   peak=-1
# )
# 
# head(bed)

write.table(peaks.df,
            paste(prefix, 'narrowPeak', sep = "."),
            sep = "\t", quote = F, row.names = F, col.names = F)
```


## Visualize Around TSS
```{r, warning=FALSE}
binOverFeature(peaks.gr, annotationData=geneCode.tr.gr,
               # select = "nearest",
               # PeakLocForDistance="middle",
               # featureSite="FeatureStart",
               radius=10000, nbins=500, FUN=length, 
               errFun=0,
               ylab="count", 
               main="Distribution of aggregated peak numbers around TSS")

binOverFeature(peaks.gr, annotationData=geneCode.tr.gr,
               # select = "nearest",
               # PeakLocForDistance="middle",
               # featureSite="FeatureStart",
               radius=5000, nbins=500, FUN=length, 
               errFun=0,
               ylab="count", 
               main="Distribution of aggregated peak numbers around TSS")
```

## Annotation Summary
- Peak centric view

```{r, warning=F}
aCR<-assignChromosomeRegion(peaks.gr, 
                            nucleotideLevel=F, # peak centric view
                            precedence=c("Promoters", 
                                         "fiveUTRs", 
                                         "threeUTRs", 
                                         "Exons", 
                                         "immediateDownstream",  
                                         "Introns"), # Rui's precedence
                            proximal.promoter.cutoff	<- 3000,
                            immediate.downstream.cutoff	<- 1000, 
                            TxDb=TxDb)

op <- par(mar=c(11,4,4,2)) # allows the names.arg below the barplot
barplot(aCR$percentage, las=3)
rm(op)
```


## Annotate to TSS (at transcript level, geneCode.tr.gr)
- EnsDb.gr Works, but Peak1 anno wrong
- use geneCode.tr.gr
```{r, warning=F}
# Pie about annotation
# seqlevelsStyle(peaks.gr) <- seqlevelsStyle(TxDb.gr)
# seqlevelsStyle(peaks.gr) <- seqlevelsStyle(EnsDb.gr)
# todo: strand issue?
gr.anno <- annotatePeakInBatch(myPeakList = peaks.gr, 
                               AnnotationData =  geneCode.tr.gr, 
                               FeatureLocForDistance="TSS",
                               bindingRegion = c(-3000, +3000),
                               output = "overlapping", 
                               select = 'all',
                               ignore.strand = T
)
#table(gr.anno$insideFeature)
barplot(table(gr.anno$insideFeature))
#head(data.frame(gr.anno))
```

## Output full duplicated annotation (unannotated skipped, same gene name considered same)
```{r}
gr.anno.df <- data.frame(gr.anno)

# select columns
gr.anno.df <- gr.anno.df[, c("seqnames", "start", "end", "peak_id", "score", "strand", "signalValue", "pValue", "qValue", "width",
                             "geneName", "geneID", "geneBiotype")] # affects function collapse_anno

# Delete duplicated rows (same peak, same gene)
gr.anno.df$dedup_id <- paste(gr.anno.df$peak_id,gr.anno.df$geneName)
gr.anno.df<-gr.anno.df %>% distinct(dedup_id, .keep_all = TRUE)
# write.csv(gr.anno.df, 
#           paste(prefix, 'anno.WithDup.csv', sep = "."),
#           row.names = F)
WriteXLS(x = gr.anno.df,
         ExcelFileName = paste(prefix, 'anno.WithDup.xlsx', sep = "."),
         row.names = F, SheetNames = 'sheet1', na = '-')  # for user

gr.anno.df.protein_coding <- subset(gr.anno.df, geneBiotype=='protein_coding')
# write.csv(gr.anno.df.protein_coding, 
#           paste(prefix, 'anno.WithDup.protein_coding.csv', sep = "."),
#           row.names = F)
WriteXLS(x = gr.anno.df.protein_coding,
         ExcelFileName = paste(prefix, 'anno.WithDup.protein_coding.xlsx', sep = "."),
         row.names = F, SheetNames = 'sheet1', na = '-')  # for user
```
# Collapse by peak_id
```{r}
RM <- function(x) gsub(",.*", "", x)

collapse_anno <- function(gr.anno.df, outname){
  # collapse rows by peak_id
  gr.anno.df.collapsed <- gr.anno.df %>%
    group_by(peak_id) %>%
    summarise_each(funs(toString))
  
  # not collapse seqnames, start, end, width, strand, conc ...
  gr.anno.df.collapsed <- data.frame(apply(gr.anno.df.collapsed[,1:10], 2, RM), 
                                     gr.anno.df.collapsed[,11:14])  # number of columns matters
  
  print(outname)
  print(dim(gr.anno.df.collapsed))
  #write.csv(gr.anno.df.collapsed, outname, row.names = F)
  WriteXLS(x = gr.anno.df.collapsed,
           ExcelFileName = outname,
           row.names = F, SheetNames = 'sheet1', na = '-')  # for user
  
  return(gr.anno.df.collapsed)
}


gr.anno.df.collapsed <- collapse_anno(gr.anno.df, 
               paste(prefix, "anno.collapsed.xlsx", sep = "."))

gr.anno.df.protein_coding.collapsed <- collapse_anno(gr.anno.df.protein_coding, 
              paste(prefix, "anno.protein_coding.collapsed.xlsx", sep = "."))
```

# Full talbe including unannotated
```{r}
full.collapsed.df <- merge(peaks.df, gr.anno.df.collapsed[, c(1, 11:14)], 
                           by= "peak_id", all.x=T)
#full.collapsed.df <- full.collapsed.df[order(nchar(full.collapsed.df$peak_id)),]
print(dim(full.collapsed.df))
head(full.collapsed.df)
#write.csv(full.collapsed.df, "full_anno.csv", row.names = F)
WriteXLS(x = full.collapsed.df,
         ExcelFileName = paste(prefix, "full_anno.xlsx", sep = "."), 
         row.names = F, SheetNames = 'sheet1', na = '-')  # for user
```


## Enrichment
### Parameters
```{r}
library("org.Hs.eg.db")
#library("org.Mm.eg.db")
#library('reactome.db')
library("KEGG.db")
org_eg_eb <- "org.Hs.eg.db"
max_BH = 0.1
```

```{r}
keys <- unique(gr.anno.df$geneName)
```

### GO
```{r}
paste(org_eg_eb)

enriched.GO.both.condense <- getEnrichedGO(
    keys, 
    orgAnn=org_eg_eb, 
    feature_id_type = "gene_symbol", 
    maxP=max_BH,
    minGOterm=10,
    multiAdjMethod= "BH", 
    condense = T  # all genes in one line
)

```

### KEGG
```{r}
kegg <- getEnrichedPATH(
    keys, 
    orgAnn=org_eg_eb, 
    pathAnn = "KEGG.db",
    feature_id_type = "gene_symbol", 
    maxP=max_BH,
    minPATHterm=10,
    multiAdjMethod= "BH"
)

kegg <- condenseMatrixByColnames(as.matrix(kegg), "path.id")

#head(kegg)

```


### Output
```{r}
paste("enriched.GO.both.condense$bp", dim(enriched.GO.both.condense$bp)[1])
WriteXLS(x = enriched.GO.both.condense$bp,
        ExcelFileName = paste('GO_biological_process', max_BH, 'xlsx', sep = '.'), 
        row.names = F, SheetNames = 'sheet1', na = '-')

paste("enriched.GO.both.condense$cc", dim(enriched.GO.both.condense$cc)[1])
WriteXLS(x = enriched.GO.both.condense$cc,
        ExcelFileName = paste('GO_cellular_component', max_BH, 'xlsx', sep = '.'),
        row.names = F, SheetNames = 'sheet1', na = '-')

paste("enriched.GO.both.condense$mf", dim(enriched.GO.both.condense$mf)[1])
WriteXLS(x = enriched.GO.both.condense$mf,
        ExcelFileName = paste('GO_mulecular_function', max_BH, 'xlsx', sep = '.'),
        row.names = F, SheetNames = 'sheet1', na = '-')

# WriteXLS(x = enriched.PATH.both.condense,
#         ExcelFileName = 'reactome.db.FDR0.05.FC1.5.xlsx', row.names = F, SheetNames = 'sheet1', na = '-')

paste("kegg", dim(kegg)[1])
WriteXLS(x = kegg,
        ExcelFileName = paste('KEGG', max_BH, 'xlsx', sep = '.'),
        row.names = F, SheetNames = 'sheet1', na = '-')

```



### GO wordCloud plot
```{r fig.width=20, fig.height=20}
# install.packages("wordcloud")
library("wordcloud")
wordcloudPlot <- function(enrich.df, max.words = 20){
  word0 = as.character.factor(enrich.df$go.term)
  word = unlist(lapply(word0, RM))  # remove comma and after
  pvalue = enrich.df$pvalue

  df <- data.frame(word0, word, pvalue)
  df <- df[order(df$pvalue), ]
  if (max.words < dim(df)[1]) df <- df[1:max.words, ]
  
  df$num <- rank(-df$pvalue) # small pvalue to large number
  
  wordcloud(df$word, df$num, 
        # max.words = max.words, 
         scale=c(4,1) ,
        #colors = rainbow(max.words),
         #colors = brewer.pal(n=11,name="dark")
        colors = c("Red", "orange", "gold", "green", "turquoise1", "blue", "purple" ), 
         #random.order = F, random.color = F,
         )
  
  return(df)
}
```

### Biological Process
```{r fig.width=20, fig.height=20}
wordCloud <- wordcloudPlot(enriched.GO.both.condense$bp, max.words = 30)
```

### Molecular Function
```{r fig.width=20, fig.height=20}
wordCloud <- wordcloudPlot(enriched.GO.both.condense$mf, max.words = 30)
```

### Cellular Component
```{r fig.width=20, fig.height=20}
wordCloud <- wordcloudPlot(enriched.GO.both.condense$cc, max.words = 30)
```

### KEGG
```{r fig.width=20, fig.height=20}
colnames(kegg)[9] <- "go.term"
wordCloud <- wordcloudPlot(kegg, max.words = 30)
```







<!-- ```{r} -->
<!-- #install.packages("GOplot") -->
<!-- library(GOplot) -->
<!-- GOBar(subset(circ, category == 'BP')) -->
<!-- GOBubble(circ, labels = 4) -->

<!-- ``` -->




## print versions
```{r}
sessionInfo()
```